runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaJobInvocation.java

```diff
@@ -0,0 +1,122 @@
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.DONE;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.FAILED;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.RUNNING;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.STARTING;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.STOPPED;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.UNRECOGNIZED;
+import static org.apache.beam.model.jobmanagement.v1.JobApi.JobState.Enum.UPDATED;
+
+import java.util.function.Consumer;
+import org.apache.beam.model.jobmanagement.v1.JobApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.core.construction.graph.GreedyPipelineFuser;
+import org.apache.beam.runners.fnexecution.jobsubmission.JobInvocation;
+import org.apache.beam.runners.samza.util.PortablePipelineDotRenderer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Invocation of a Samza job via {@link SamzaRunner}. */
+public class SamzaJobInvocation implements JobInvocation {
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaJobInvocation.class);
+
+  private final SamzaPipelineOptions options;
+  private final RunnerApi.Pipeline originalPipeline;
+  private volatile SamzaPipelineResult pipelineResult;
+
+  public SamzaJobInvocation(RunnerApi.Pipeline pipeline, SamzaPipelineOptions options) {
+    this.originalPipeline = pipeline;
+    this.options = options;
+  }
+
+  private SamzaPipelineResult invokeSamzaJob() {
+    // Fused pipeline proto.
+    final RunnerApi.Pipeline fusedPipeline =
+        GreedyPipelineFuser.fuse(originalPipeline).toPipeline();
+    LOG.info("Portable pipeline to run:");
+    LOG.info(PortablePipelineDotRenderer.toDotString(fusedPipeline));
+    // the pipeline option coming from sdk will set the sdk specific runner which will break serialization
+    // so we need to reset the runner here to a valid Java runner
+    options.setRunner(SamzaRunner.class);
+    try {
+      final SamzaRunner runner = new SamzaRunner(options);
+      return runner.runPortablePipeline(fusedPipeline);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to invoke samza job", e);
+    }
+  }
+
+  @Override
+  public void start() {
+    LOG.info("Starting job invocation {}", getId());
+    pipelineResult = invokeSamzaJob();
+  }
+
+  @Override
+  public String getId() {
+    return options.getJobName();
+  }
+
+  @Override
+  public void cancel() {
+    try {
+      if (pipelineResult != null) {
+        pipelineResult.cancel();
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to cancel job.", e);
+    }
+  }
+
+  @Override
+  public JobApi.JobState.Enum getState() {
+    if (pipelineResult == null) {
+      return STARTING;
+    }
+    switch (pipelineResult.getState()) {
```

Comment:
Seems we can directly return pipelineResult here. Any idea why doing this switch?

Reply:
because one if the Java PipelineResult.State, the other one is the protobuf version of JobApi.JobState. That's very unfortunate.